### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by pavank-v**

### CODE REVIEW

File: Backend/base/settings.py [hunk 1]
```diff
@@ -129,6 +129,26 @@
     "REFRESH_TOKEN_LIFETIME": timedelta(days=1),
 }
 
-# for cors headers
+# for cors headers (for React integration in the future)
 CORS_ALLOW_ALL_ORIGINS = True
```
Comment (praveen-pb) [inline Backend/base/settings.py:135]: Read about how to allow specific origins instead of all. Allow all origins and allow credentials won't work together.

File: Backend/base/urls.py [hunk 2]
```diff
@@ -6,9 +6,12 @@
 
 urlpatterns = [
     path('admin/', admin.site.urls),
-    path('register/', UserRegisterView.as_view(), name='register'),
+    # Endpoints for Authentication
+    path('user/', UserRegisterView.as_view(), name='register'),
+    path('api-auth/', include('rest_framework.urls')),
+    # Endpoints for jwt
     path("token/", TokenObtainPairView.as_view(), name="get_token"),
```
Comment (praveen-pb) [inline Backend/base/urls.py]: Try to maintain separate app for authentication. We have multiple functionality in authentication. so always have it as separate app.

File: Backend/myfitapp/models.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
 from django.db import models
 from django.contrib.auth.models import User
+from django.utils import timezone
+import logging
 
-from .scripts import calories_finder
+from myfitapp.scripts import calculate_end_date, calories_finder
 
 # Global Variables
 LEVEL_CHOICES = [
```
Comment (praveen-pb) [inline Backend/myfitapp/models.py]: If you are using it in multiple places, put it as Enum or if you are using in models, use models.TextChoices.
Read about both Enum and models.TextChoices

File: Backend/myfitapp/models.py [hunk 4]
```diff
@@ -148,11 +152,15 @@ def __str__(self):
...
+    fats = models.IntegerField()
     start_date = models.DateField(auto_now_add=True)
     end_date = models.DateField(blank=True, null=True)
 
+    # Method to calculate the calorie budget for that day
     def calculate_calorie_budget(self):
-        weight = self.profile.weight
+        weight = float(self.profile.weight)
         goal = self.profile.goal
         if goal == "Lose":
```
Comment (praveen-pb) [inline Backend/myfitapp/models.py]: Lose should be models.TextChoice it will be easy to use in multiple places. don't use string values.

File: Backend/myfitapp/models.py [hunk 5]
```diff
@@ -161,8 +169,36 @@ def calculate_calorie_budget(self):
         else:  # Maintain
             return round(weight * 2.202 * 15)
 
+    # Method to calculate the macro nutrients required for an user
+    def calculate_macros(self):
+        calories = self.calorie_budget
+        if self.profile.gender == "Male":
```
Comment (praveen-pb) [inline Backend/myfitapp/models.py]: Same here, don't use string

File: Backend/myfitapp/models.py [hunk 6]
```diff
@@ -161,8 +169,36 @@ def calculate_calorie_budget(self):
...
+
     def save(self, *args, **kwargs):
+        if not self.start_date:
+            self.start_date = timezone.now().date()
+        logger.debug(f"Start Date: {self.start_date}")
         self.calorie_budget = self.calculate_calorie_budget()
+        self.protein, self.carbs, self.fats = self.calculate_macros()
+        self.end_date = calculate_end_date(self.start_date)
+        """
+        TODO: figure out this -> tried self.save() with different method name
```
Comment (praveen-pb) [inline Backend/myfitapp/models.py]: Reason is the function name. You have over-written save function save function, so this will be called first. then you call parent save, which will again save the object. check (your_model).objects.create(). this will create object with default values. save is used only when updating and to trigger signals.

File: Backend/myfitapp/models.py [hunk 7]
```diff
@@ -189,15 +225,64 @@ def save(self, *args, **kwargs):  # Overriding the save()
...
         self.carbs_intake = nutrition["carbohydrates_total_g"]
         self.fat_intake = nutrition["fat_total_g"]
+        """
+        TODO: figure out this -> tried self.save() with different method name
+          causing infinite recursion
+        """
+        super().save(*args, **kwargs)
 
-        super().save()
+    @staticmethod
```
Comment (praveen-pb) [inline Backend/myfitapp/models.py]: don't use static method inside model class. If this function is not related to this model, create a separate file like handler/helper.py and add it in that.
Reason - models should contain important functions and related to it. static methods can be anywhere.

File: Backend/myfitapp/scripts.py [hunk 8]
```diff
@@ -1,10 +1,12 @@
...
+from datetime import timedelta
 import requests
 import json
 
 
 # Helper function to get the nutritional facts
 def nutritional_facts(food_name, quantity):
+    # TODO: Hide the api key
     api_url = "https://api.calorieninjas.com/v1/nutrition?query="
     api_key = "qH8EV4KUKeCOivfUjmiiiA==gxZC8l0UwQiztZXp"
```
Comment (praveen-pb) [inline Backend/myfitapp/scripts.py]: Don't push API key to GitHub. it is risky. Read about python-dotenv library. It helps you in storing secret keys

File: Backend/myfitapp/serializers.py [hunk 9]
```diff
@@ -174,15 +163,34 @@ class Meta:
...
+            "resistance_trainings_ids",
+            "cardio_trainings_ids",
+            "crossfit_trainings_ids",
+            "flexibility_trainings_ids",
+            "recoveries_ids",
+        ]
+
     def create(self, validated_data):
-        # Separate out the IDs to assign ManyToMany relationships
+        # Extract the IDs for ManyToMany relationships
```
Comment (praveen-pb) [inline Backend/myfitapp/serializers.py]: Try to avoid using many to many relationship. that makes the code complex and difficult to maintain in long term.

File: Backend/myfitapp/serializers.py [hunk 10]
```diff
@@ -199,21 +207,48 @@ def create(self, validated_data):
...
+
+        return data
 
     def create(self, validated_data):
         profile = validated_data["profile"]
-        weight = profile.weight
-        goal = profile.goal
 
-        if goal == "Lose":
-            validated_data["calorie_budget"] = round(weight * 2.202 * 13.5)
```
Comment (praveen-pb) [inline Backend/myfitapp/serializers.py]: Add comment why the value is 2.202. it will help others to understand while reading code.

File: Backend/myfitapp/views.py [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+from myfitapp.serializers import (BodyMetricsSerializer, RecipeSerializer, 
+    CardioTrainingSerializer, CrossFitTrainingSerializer,
+    FlexibilityTrainingSerializer, SearchSerializer, RecoverySerializer,
+    PersonalWorkoutPlanSerializer,  UserSerializer, UserFoodLogSerializer,
+    ResistanceTrainingSerializer, PersonalDietPlanSerializer,
+)
+
+
+# View to Register the User
+class UserRegisterView(
```
Comment (praveen-pb) [inline Backend/myfitapp/views.py]: Use APIView. this will take care most of the thing. No need to inherit multiple mixins

File: Backend/myfitapp/views.py [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+
+        response.data = {
+            "Message": "User created successfully!!!",
+            "User-ID": response.data.get("id"),
+            "Profile-data": response.data.get("profile"),
+            "Additional-info": "Thanks for signing up",
+        }
+        return response
+
+    def post(self, request, *args, **kwargs):
```
Comment (praveen-pb) [inline Backend/myfitapp/views.py]: Try to use serializers. Serializers takes care of validations and CRUD operations easily.

File: Backend/myfitapp/views.py [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+class BodyMetricsUpdateView(generics.RetrieveUpdateAPIView):
+    """
+    With this view user can update the body metrics to
+    keep track of their progress
+    """
+
+    permission_classes = [IsAuthenticated]
+    serializer_class = BodyMetricsSerializer
+
+    def get_object(self):
```
Comment (praveen-pb) [inline Backend/myfitapp/views.py]: get_object and get methods are different. read about the differences.

File: Backend/myfitapp/views.py [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+                {"error": "Profile not found for the user."},
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        if input_serializer.is_valid():
+            input_serializer.save(profile=profile)
+
+            daily_summary = UserFoodLog.daily_summary(profile)
+
+            return Response(
```
Comment (praveen-pb) [inline Backend/myfitapp/views.py:111]: use serializers for response.

**Review by praveen-pb — APPROVED** [review]
Commented in multiple places for your learnings. Add test cases. test cases are very important.

File: Backend/myfitapp/views.py (hunk 11 above)
Comment (pavank-v) [reply Backend/myfitapp/views.py]: used mixins for the html form, try to change it to APIView

File: Backend/myfitapp/views.py (hunk 13 above)
Comment (pavank-v) [reply Backend/myfitapp/views.py]: I have read the difference. Thanks for mentioning

File: Backend/myfitapp/views.py (hunk 14 above)
Comment (pavank-v) [reply Backend/myfitapp/views.py:111]: did this for customized response.
changed to serializers now.

File: Backend/myfitapp/models.py (hunk 3 above)
Comment (pavank-v) [reply Backend/myfitapp/models.py]: Done

File: Backend/myfitapp/models.py (hunk 6 above)
Comment (pavank-v) [reply Backend/myfitapp/models.py]: got it

File: Backend/myfitapp/models.py (hunk 7 above)
Comment (pavank-v) [reply Backend/myfitapp/models.py]: done

File: Backend/myfitapp/models.py (hunk 4 above)
Comment (pavank-v) [reply Backend/myfitapp/models.py]: done

File: Backend/myfitapp/scripts.py (hunk 8 above)
Comment (pavank-v) [reply Backend/myfitapp/scripts.py]: done

File: Backend/myfitapp/serializers.py (hunk 9 above)
Comment (pavank-v) [reply Backend/myfitapp/serializers.py]: okay
